system/jlib/jhash.hpp

```diff
@@ -687,4 +688,104 @@ class CTimeLimitedCache
     }
 };
 
+/**
+ * CLRUCache
+ *
+ * Least-Recently-Used cache class, specialized for key and
+ * value pointer types (the value is a pointer or a data type
+ * where a nullptr could represent a missing value).
+ *
+ * The get() method returns a found object by value.  This
+ * is intentional and very useful for maintaining refcounts.
+ *
+ * There is a minimum size for the cache, defined by
+ * LRU_MIN_CACHE_SIZE.  Attempts to create a smaller cache
+ * will be silently changed to the minimum size.  If no
+ * initial size is provided to the constructor, the cache
+ * size will be set to LRU_MIN_CACHE_SIZE.
+ *
+ * Methods here are not thread-safe.  Callers should block
+ * concurrent access for non-const methods (which are most
+ * of them).
+ */
+
+#define LRU_MIN_CACHE_SIZE 10
+
+template <class KEYTYPE, class PTRTYPE>
+class CLRUCache
+{
+    private:
+        std::list<KEYTYPE> recentList;
+        std::unordered_map<KEYTYPE, std::pair<PTRTYPE, typename std::list<KEYTYPE>::iterator>> lookupMap;
+        size32_t maxCacheSize;
+
+        void _downsize()
+        {
+            while (lookupMap.size() > maxCacheSize)
+            {
+                lookupMap.erase(recentList.back());
+                recentList.pop_back();
+            }
+        }
+    
+    public:
+        CLRUCache() : maxCacheSize(LRU_MIN_CACHE_SIZE) {}
+        CLRUCache(size32_t _maxCacheSize) : maxCacheSize(_maxCacheSize < LRU_MIN_CACHE_SIZE ? LRU_MIN_CACHE_SIZE : _maxCacheSize) {}
+        CLRUCache(const CLRUCache& other) = delete;
+        ~CLRUCache() = default;
+
+        size32_t getCacheSize() const
+        {
+            return lookupMap.size();
+        }
+
+        size32_t setMaxCacheSize(size32_t _maxCacheSize)
+        {
+            maxCacheSize = _maxCacheSize < LRU_MIN_CACHE_SIZE ? LRU_MIN_CACHE_SIZE : _maxCacheSize;
+            _downsize();
+            return maxCacheSize;
+        }
+
+        PTRTYPE get(const KEYTYPE& key)
+        {
+            auto foundIter = lookupMap.find(key);
+            if (foundIter == lookupMap.end())
+                return nullptr;
+            
+            recentList.erase(foundIter->second.second);
+            recentList.push_front(key);
+            lookupMap[key] = {foundIter->second.first, recentList.begin()};
+            return foundIter->second.first;
+        }
+
+        void set(const KEYTYPE& key, const PTRTYPE& value)
+        {
+            auto foundIter = lookupMap.find(key);
+            
+            if (foundIter == lookupMap.end())
+            {
+                recentList.push_front(key);
+                lookupMap[key] = {value, recentList.begin()};
+                _downsize();
+            }
+            else
+            {
+                recentList.erase(foundIter->second.second);
+                recentList.push_front(key);
+                lookupMap[key] = {value, recentList.begin()};
```

Comment:
foundIter->second = ...

is probably more efficient, but it shouldn't be hit in the current usage

Reply:
Fixed.

Agreed that in the current regex scenario this block won't be hit.  I really struggled with creating a general purpose feature at this level of code versus just what I needed, and reigning in my desire to add another version of this LRU cache that would be more efficient for non-pointer objects.